plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/CompoundProcessor.java

```diff
@@ -59,23 +62,32 @@ public String getType() {
 
     @Override
     public void execute(IngestDocument ingestDocument) throws Exception {
-        try {
-            for (Processor processor : processors) {
+        for (Processor processor : processors) {
+            try {
                 processor.execute(ingestDocument);
-            }
-        } catch (Exception e) {
-            if (onFailureProcessors.isEmpty()) {
-                throw e;
-            } else {
-                executeOnFailure(ingestDocument);
+            } catch (Exception e) {
+                if (onFailureProcessors.isEmpty()) {
+                    throw e;
+                } else {
+                    executeOnFailure(ingestDocument, e, processor.getType());
+                }
+                return;
             }
         }
+    }
+
+    private void resetErrorContext(Map<String, String> ingestMetadata) {
 
     }
 
-    void executeOnFailure(IngestDocument ingestDocument) throws Exception {
+    void executeOnFailure(IngestDocument ingestDocument, Exception cause, String failedProcessorType) throws Exception {
+        Map<String, String> ingestMetadata = ingestDocument.getIngestMetadata();
+        ingestMetadata.put(ON_FAILURE_MESSAGE_FIELD, cause.getMessage());
+        ingestMetadata.put(ON_FAILURE_PROCESSOR_FIELD, failedProcessorType);
```

Comment:
I agree. I was hoping to include a more complex object with subfields that can be probed during pipeline handling..  but the metadata is of type `String -> String`, so I figured this would be the best can be done within that limitation. Should be sufficient for now.